// @circle-fin/x402-batching/dist/client/index.mjs
// src/client/BatchEvmScheme.ts
import { getAddress } from "viem";

// src/constants.ts
var CIRCLE_BATCHING_NAME = "GatewayWalletBatched";
var CIRCLE_BATCHING_VERSION = "1";
var CIRCLE_BATCHING_SCHEME = "exact";
var GATEWAY_MIN_AUTH_VALIDITY_SECONDS = 7 * 24 * 60 * 60;
var GATEWAY_AUTH_VALIDITY_BUFFER_SECONDS = 100;
var GATEWAY_AUTH_VALIDITY_WINDOW_SECONDS = GATEWAY_MIN_AUTH_VALIDITY_SECONDS + GATEWAY_AUTH_VALIDITY_BUFFER_SECONDS;
var ARC_PRIVATE_MAINNET_HEADER = "X-ARC-PRIVATE-MAINNET-ENABLED";
function arcPrivateMainnetHeaders(arcPrivateMainnet) {
  if (!arcPrivateMainnet) {
    return {};
  }
  return { [ARC_PRIVATE_MAINNET_HEADER]: "true" };
}

// src/detection.ts
function supportsBatching(requirements) {
  const extra = requirements.extra;
  if (!extra) {
    return false;
  }
  return extra.name === CIRCLE_BATCHING_NAME && extra.version === CIRCLE_BATCHING_VERSION;
}
function getVerifyingContract(requirements) {
  if (!supportsBatching(requirements)) {
    return void 0;
  }
  const verifyingContract = requirements.extra?.verifyingContract;
  if (typeof verifyingContract === "string") {
    return verifyingContract;
  }
  return void 0;
}

// src/hooks.ts
function warnHookFailure(component, phase, label, error) {
  const detail = error instanceof Error ? error.message : String(error);
  console.warn(`[x402] ${component} ${phase} hook threw (${label}): ${detail}`);
}

// src/client/BatchEvmScheme.ts
var authorizationTypes = {
  TransferWithAuthorization: [
    { name: "from", type: "address" },
    { name: "to", type: "address" },
    { name: "value", type: "uint256" },
    { name: "validAfter", type: "uint256" },
    { name: "validBefore", type: "uint256" },
    { name: "nonce", type: "bytes32" }
  ]
};
function createNonce() {
  const bytes = new Uint8Array(32);
  crypto.getRandomValues(bytes);
  return `0x${Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("")}`;
}
var BatchEvmScheme = class {
  /**
   * Creates a new BatchEvmScheme.
   *
   * @param signer - The EVM signer for signing payment authorizations
   */
  constructor(signer) {
    this.signer = signer;
  }
  scheme = CIRCLE_BATCHING_SCHEME;
  beforePaymentCreationHooks = [];
  afterPaymentCreationHooks = [];
  onPaymentCreationFailureHooks = [];
  paymentResponseHooks = [];
  /**
   * Register a hook that runs before payment payload creation. Returning
   * `{ abort: true, reason }` causes `createPaymentPayload` to throw
   * `Error("Payment creation aborted: <reason>")`.
   */
  onBeforePaymentCreation(hook) {
    this.beforePaymentCreationHooks.push(hook);
    return this;
  }
  /**
   * Register a hook that runs after a successful payment payload creation.
   */
  onAfterPaymentCreation(hook) {
    this.afterPaymentCreationHooks.push(hook);
    return this;
  }
  /**
   * Register a hook that runs when payload creation throws. Return
   * `{ recovered: true, payload }` to swallow the error and use the
   * provided payload as the result instead.
   */
  onPaymentCreationFailure(hook) {
    this.onPaymentCreationFailureHooks.push(hook);
    return this;
  }
  /**
   * Register a hook that runs after the buyer's paid HTTP request completes.
   * Hooks fire when {@link dispatchPaymentResponse} is called by the
   * transport. Return `{ recovered: true }` to instruct the transport to
   * retry with a fresh payload.
   */
  onPaymentResponse(hook) {
    this.paymentResponseHooks.push(hook);
    return this;
  }
  /**
   * Fire all registered `onPaymentResponse` hooks in order. Returns
   * `{ recovered: true }` if any hook signals recovery (first wins).
   *
   * Transport wrappers should call this after the paid HTTP request settles
   * — see `x402HTTPClient.processPaymentResult` in the standard SDK.
   */
  async dispatchPaymentResponse(context) {
    for (const [index, hook] of this.paymentResponseHooks.entries()) {
      try {
        const result = await hook(context);
        if (result && "recovered" in result && result.recovered) {
          return { recovered: true };
        }
      } catch (err) {
        warnHookFailure(
          "BatchEvmScheme",
          "onPaymentResponse",
          `manual hook #${index}`,
          err
        );
      }
    }
    return void 0;
  }
  /**
   * Creates a payment payload for Circle batching.
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements (must be a Circle batching option)
   * @returns Promise resolving to the payment payload
   * @throws Error if requirements are not a Circle batching option
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const context = {
      paymentRequired: this.synthesizePaymentRequired(x402Version, paymentRequirements),
      selectedRequirements: paymentRequirements
    };
    for (const hook of this.beforePaymentCreationHooks) {
      const result = await hook(context);
      if (result && "abort" in result && result.abort) {
        throw new Error(`Payment creation aborted: ${result.reason}`);
      }
    }
    try {
      if (!supportsBatching(paymentRequirements)) {
        throw new Error(
          `BatchEvmScheme can only handle Circle batching options. Expected extra.name="${CIRCLE_BATCHING_NAME}" and extra.version="${CIRCLE_BATCHING_VERSION}"`
        );
      }
      const verifyingContract = getVerifyingContract(paymentRequirements);
      if (!verifyingContract) {
        throw new Error(
          "Circle batching option missing extra.verifyingContract (GatewayWallet address)"
        );
      }
      const nonce = createNonce();
      const now = Math.floor(Date.now() / 1e3);
      const validityWindowSeconds = Math.max(
        paymentRequirements.maxTimeoutSeconds,
        GATEWAY_AUTH_VALIDITY_WINDOW_SECONDS
      );
      const authorization = {
        from: this.signer.address,
        to: getAddress(paymentRequirements.payTo),
        value: paymentRequirements.amount,
        validAfter: (now - 600).toString(),
        // 10 minutes before
        // Clamp seller-published timeouts so the signed authorization is always
        // valid long enough for Gateway batching.
        validBefore: (now + validityWindowSeconds).toString(),
        nonce
      };
      const signature = await this.signAuthorization(
        authorization,
        paymentRequirements,
        verifyingContract
      );
      const payload = {
        authorization,
        signature
      };
      const partial = {
        x402Version,
        payload
      };
      const created = {
        x402Version,
        payload
      };
      const createdContext = {
        ...context,
        paymentPayload: created
      };
      for (const [index, hook] of this.afterPaymentCreationHooks.entries()) {
        try {
          await hook(createdContext);
        } catch (err) {
          warnHookFailure(
            "BatchEvmScheme",
            "afterPaymentCreation",
            `manual hook #${index}`,
            err
          );
        }
      }
      return partial;
    } catch (error) {
      const failureContext = {
        ...context,
        error
      };
      for (const hook of this.onPaymentCreationFailureHooks) {
        const result = await hook(failureContext);
        if (result && "recovered" in result && result.recovered) {
          return {
            x402Version: result.payload.x402Version,
            payload: result.payload.payload
          };
        }
      }
      throw error;
    }
  }
  /**
   * Sign the EIP-3009 authorization using EIP-712.
   * Uses the GatewayWallet contract as verifyingContract instead of USDC.
   */
  async signAuthorization(authorization, requirements, verifyingContract) {
    if (!requirements.network.startsWith("eip155:")) {
      throw new Error(
        `BatchEvmScheme: unsupported network format "${requirements.network}". Expected "eip155:<chainId>"`
      );
    }
    const chainId = parseInt(requirements.network.split(":")[1]);
    const domain = {
      name: CIRCLE_BATCHING_NAME,
      version: CIRCLE_BATCHING_VERSION,
      chainId,
      verifyingContract: getAddress(verifyingContract)
    };
    const message = {
      from: getAddress(authorization.from),
      to: getAddress(authorization.to),
      value: BigInt(authorization.value),
      validAfter: BigInt(authorization.validAfter),
      validBefore: BigInt(authorization.validBefore),
      nonce: authorization.nonce
    };
    return await this.signer.signTypedData({
      domain,
      types: authorizationTypes,
      primaryType: "TransferWithAuthorization",
      message
    });
  }
  /**
   * When `createPaymentPayload` is called directly (not through an x402Client),
   * we don't have a real PaymentRequired response. Synthesize a minimal one
   * so the hook context still carries the standard SDK's shape.
   */
  synthesizePaymentRequired(x402Version, requirements) {
    return {
      x402Version,
      accepts: [requirements]
    };
  }
};

// src/client/CompositeEvmScheme.ts
var CompositeEvmScheme = class {
  constructor(batchScheme, fallbackScheme) {
    this.batchScheme = batchScheme;
    this.fallbackScheme = fallbackScheme;
    if (fallbackScheme.scheme !== CIRCLE_BATCHING_SCHEME) {
      throw new Error(
        `CompositeEvmScheme: fallbackScheme.scheme must be "${CIRCLE_BATCHING_SCHEME}", got "${fallbackScheme.scheme}"`
      );
    }
  }
  scheme = CIRCLE_BATCHING_SCHEME;
  async createPaymentPayload(x402Version, paymentRequirements) {
    if (supportsBatching(paymentRequirements)) {
      return this.batchScheme.createPaymentPayload(x402Version, paymentRequirements);
    }
    return this.fallbackScheme.createPaymentPayload(x402Version, paymentRequirements);
  }
};

// src/client/register.ts
function registerBatchScheme(client, config) {
  const batchScheme = new BatchEvmScheme(config.signer);
  const schemeToRegister = config.fallbackScheme ? new CompositeEvmScheme(batchScheme, config.fallbackScheme) : batchScheme;
  const networks = config.networks ?? ["eip155:*"];
  for (const network of networks) {
    client.register(network, schemeToRegister);
  }
  return schemeToRegister;
}

// src/client/GatewayClient.ts
import {
  createPublicClient,
  createWalletClient,
  http,
  parseUnits,
  formatUnits,
  pad,
  zeroAddress,
  maxUint256,
  defineChain,
  erc20Abi
} from "viem";
import { privateKeyToAccount } from "viem/accounts";
import * as chains from "viem/chains";
import { randomBytes } from "crypto";
var arcMainnet = chains.arc ?? defineChain({
  id: 5042,
  name: "Arc",
  nativeCurrency: { decimals: 18, name: "USDC", symbol: "USDC" },
  rpcUrls: {
    default: { http: [] }
  }
});
var arcTestnetChain = chains.arcTestnet ?? defineChain({
  id: 5042002,
  name: "Arc Testnet",
  nativeCurrency: { decimals: 18, name: "USDC", symbol: "USDC" },
  rpcUrls: {
    default: { http: ["https://rpc.testnet.arc.network"] }
  },
  testnet: true
});
var sonicTestnet = defineChain({
  id: 14601,
  name: "Sonic Testnet",
  nativeCurrency: { decimals: 18, name: "Sonic", symbol: "S" },
  rpcUrls: {
    default: { http: ["https://rpc.testnet.soniclabs.com"] }
  },
  blockExplorers: {
    default: { name: "Sonic Testnet Explorer", url: "https://testnet.soniclabs.com/" }
  },
  testnet: true
});
var GATEWAY_API_TESTNET = "https://gateway-api-testnet.circle.com/v1";
var GATEWAY_API_MAINNET = "https://gateway-api.circle.com/v1";
var GATEWAY_MINTER_ABI = [
  {
    name: "gatewayMint",
    type: "function",
    stateMutability: "nonpayable",
    inputs: [
      { name: "attestationPayload", type: "bytes" },
      { name: "signature", type: "bytes" }
    ],
    outputs: []
  }
];
var GATEWAY_WALLET_ABI = [
  {
    name: "deposit",
    type: "function",
    stateMutability: "nonpayable",
    inputs: [
      { name: "token", type: "address" },
      { name: "value", type: "uint256" }
    ],
    outputs: []
  },
  {
    name: "depositFor",
    type: "function",
    stateMutability: "nonpayable",
    inputs: [
      { name: "token", type: "address" },
      { name: "depositor", type: "address" },
      { name: "value", type: "uint256" }
    ],
    outputs: []
  },
  {
    name: "totalBalance",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "token", type: "address" },
      { name: "depositor", type: "address" }
    ],
    outputs: [{ name: "", type: "uint256" }]
  },
  {
    name: "availableBalance",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "token", type: "address" },
      { name: "depositor", type: "address" }
    ],
    outputs: [{ name: "", type: "uint256" }]
  },
  {
    name: "withdrawingBalance",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "token", type: "address" },
      { name: "depositor", type: "address" }
    ],
    outputs: [{ name: "", type: "uint256" }]
  },
  {
    name: "withdrawableBalance",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "token", type: "address" },
      { name: "depositor", type: "address" }
    ],
    outputs: [{ name: "", type: "uint256" }]
  },
  {
    name: "withdrawalDelay",
    type: "function",
    stateMutability: "view",
    inputs: [],
    outputs: [{ name: "", type: "uint256" }]
  },
  {
    name: "withdrawalBlock",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "token", type: "address" },
      { name: "depositor", type: "address" }
    ],
    outputs: [{ name: "", type: "uint256" }]
  },
  {
    name: "initiateWithdrawal",
    type: "function",
    stateMutability: "nonpayable",
    inputs: [
      { name: "token", type: "address" },
      { name: "value", type: "uint256" }
    ],
    outputs: []
  },
  {
    name: "withdraw",
    type: "function",
    stateMutability: "nonpayable",
    inputs: [{ name: "token", type: "address" }],
    outputs: []
  }
];
var GATEWAY_DOMAINS = {
  // Testnet
  arbitrumSepolia: 3,
  arcTestnet: 26,
  avalancheFuji: 1,
  baseSepolia: 6,
  sepolia: 0,
  hyperEvmTestnet: 19,
  optimismSepolia: 2,
  polygonAmoy: 7,
  seiAtlantic: 16,
  sonicTestnet: 13,
  unichainSepolia: 10,
  worldChainSepolia: 14,
  // Mainnet
  arbitrum: 3,
  avalanche: 1,
  base: 6,
  ethereum: 0,
  hyperEvm: 19,
  optimism: 2,
  polygon: 7,
  sei: 16,
  sonic: 13,
  unichain: 10,
  worldChain: 14,
  arc: 26
};
var TESTNET_GATEWAY_WALLET = "0x0077777d7EBA4688BDeF3E311b846F25870A19B9";
var TESTNET_GATEWAY_MINTER = "0x0022222ABE238Cc2C7Bb1f21003F0a260052475B";
var MAINNET_GATEWAY_WALLET = "0x77777777Dcc4d5A8B6E418Fd04D8997ef11000eE";
var MAINNET_GATEWAY_MINTER = "0x2222222d7164433c4C09B0b0D809a9b52C04C205";
var CHAIN_CONFIGS = {
  // Testnet chains
  arbitrumSepolia: {
    chain: chains.arbitrumSepolia,
    domain: GATEWAY_DOMAINS.arbitrumSepolia,
    usdc: "0x75faf114eafb1BDbe2F0316DF893fd58CE46AA4d",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  arcTestnet: {
    chain: arcTestnetChain,
    domain: GATEWAY_DOMAINS.arcTestnet,
    usdc: "0x3600000000000000000000000000000000000000",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER,
    rpcUrl: "https://rpc.testnet.arc.network"
  },
  avalancheFuji: {
    chain: chains.avalancheFuji,
    domain: GATEWAY_DOMAINS.avalancheFuji,
    usdc: "0x5425890298aed601595a70AB815c96711a31Bc65",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  baseSepolia: {
    chain: chains.baseSepolia,
    domain: GATEWAY_DOMAINS.baseSepolia,
    usdc: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER,
    rpcUrl: "https://sepolia-preconf.base.org"
  },
  sepolia: {
    chain: chains.sepolia,
    domain: GATEWAY_DOMAINS.sepolia,
    usdc: "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  hyperEvmTestnet: {
    chain: chains.hyperliquidEvmTestnet,
    domain: GATEWAY_DOMAINS.hyperEvmTestnet,
    usdc: "0x2B3370eE501B4a559b57D449569354196457D8Ab",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  optimismSepolia: {
    chain: chains.optimismSepolia,
    domain: GATEWAY_DOMAINS.optimismSepolia,
    usdc: "0x5fd84259d66Cd46123540766Be93DFE6D43130D7",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  polygonAmoy: {
    chain: chains.polygonAmoy,
    domain: GATEWAY_DOMAINS.polygonAmoy,
    usdc: "0x41E94Eb019C0762f9Bfcf9Fb1E58725BfB0e7582",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  seiAtlantic: {
    chain: chains.seiTestnet,
    domain: GATEWAY_DOMAINS.seiAtlantic,
    usdc: "0x4fCF1784B31630811181f670Aea7A7bEF803eaED",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  sonicTestnet: {
    chain: sonicTestnet,
    domain: GATEWAY_DOMAINS.sonicTestnet,
    usdc: "0x0BA304580ee7c9a980CF72e55f5Ed2E9fd30Bc51",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  unichainSepolia: {
    chain: chains.unichainSepolia,
    domain: GATEWAY_DOMAINS.unichainSepolia,
    usdc: "0x31d0220469e10c4E71834a79b1f276d740d3768F",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  worldChainSepolia: {
    chain: chains.worldchainSepolia,
    domain: GATEWAY_DOMAINS.worldChainSepolia,
    usdc: "0x66145f38cBAC35Ca6F1Dfb4914dF98F1614aeA88",
    gatewayWallet: TESTNET_GATEWAY_WALLET,
    gatewayMinter: TESTNET_GATEWAY_MINTER
  },
  // Mainnet chains
  arbitrum: {
    chain: chains.arbitrum,
    domain: GATEWAY_DOMAINS.arbitrum,
    usdc: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  avalanche: {
    chain: chains.avalanche,
    domain: GATEWAY_DOMAINS.avalanche,
    usdc: "0xB97EF9Ef8734C71904D8002F8b6Bc66Dd9c48a6E",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  base: {
    chain: chains.base,
    domain: GATEWAY_DOMAINS.base,
    usdc: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  ethereum: {
    chain: chains.mainnet,
    domain: GATEWAY_DOMAINS.ethereum,
    usdc: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER,
    rpcUrl: "https://ethereum-rpc.publicnode.com"
  },
  hyperEvm: {
    chain: chains.hyperEvm,
    domain: GATEWAY_DOMAINS.hyperEvm,
    usdc: "0xb88339CB7199b77E23DB6E890353E22632Ba630f",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  optimism: {
    chain: chains.optimism,
    domain: GATEWAY_DOMAINS.optimism,
    usdc: "0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  polygon: {
    chain: chains.polygon,
    domain: GATEWAY_DOMAINS.polygon,
    usdc: "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER,
    rpcUrl: "https://poly.api.pocket.network"
  },
  sei: {
    chain: chains.sei,
    domain: GATEWAY_DOMAINS.sei,
    usdc: "0xe15fC38F6D8c56aF07bbCBe3BAf5708A2Bf42392",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER,
    rpcUrl: "https://sei.api.pocket.network"
  },
  sonic: {
    chain: chains.sonic,
    domain: GATEWAY_DOMAINS.sonic,
    usdc: "0x29219dd400f2Bf60E5a23d13Be72B486D4038894",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  unichain: {
    chain: chains.unichain,
    domain: GATEWAY_DOMAINS.unichain,
    usdc: "0x078D782b760474a361dDA0AF3839290b0EF57AD6",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  worldChain: {
    chain: chains.worldchain,
    domain: GATEWAY_DOMAINS.worldChain,
    usdc: "0x79A02482A880bCE3F13e09Da970dC34db4CD24d1",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  },
  arc: {
    // No `rpcUrl`: Arc mainnet has no public RPC until ~2026-06-22. Partners must
    // pass their private RPC via GatewayClientConfig.rpcUrl (enforced in the
    // GatewayClient constructor).
    chain: arcMainnet,
    domain: GATEWAY_DOMAINS.arc,
    usdc: "0x3600000000000000000000000000000000000000",
    gatewayWallet: MAINNET_GATEWAY_WALLET,
    gatewayMinter: MAINNET_GATEWAY_MINTER
  }
};
var GatewayClient = class {
  chainConfig;
  account;
  publicClient;
  walletClient;
  _arcPrivateMainnet;
  /**
   * The scheme that builds and signs the batched payment payload. `pay()`
   * delegates to this so the buyer flow runs the same lifecycle hooks as the
   * standalone `BatchEvmScheme` / `x402Client` paths.
   */
  batchScheme;
  /**
   * Creates a new GatewayClient.
   *
   * @param config - Configuration including chain and private key
   */
  constructor(config) {
    const chainConfig = CHAIN_CONFIGS[config.chain];
    if (!chainConfig) {
      throw new Error(`Unsupported chain: ${config.chain}`);
    }
    this.chainConfig = chainConfig;
    this._arcPrivateMainnet = config.arcPrivateMainnet ?? config.chain === "arc";
    this.account = privateKeyToAccount(config.privateKey);
    const rpcUrl = config.rpcUrl ?? chainConfig.rpcUrl;
    if (config.chain === "arc" && !rpcUrl) {
      throw new Error(
        "Arc mainnet has no public RPC yet. Pass a private RPC URL via `new GatewayClient({ chain: 'arc', rpcUrl: 'https://<your-private-arc-rpc>' })` (Alchemy / Quicknode / Circle-provided)."
      );
    }
    const transport = rpcUrl ? http(rpcUrl) : http();
    this.publicClient = createPublicClient({
      chain: chainConfig.chain,
      transport
    });
    this.walletClient = createWalletClient({
      account: this.account,
      chain: chainConfig.chain,
      transport
    });
    const signer = {
      address: this.account.address,
      signTypedData: (params) => this.account.signTypedData(params)
    };
    this.batchScheme = new BatchEvmScheme(signer);
  }
  // ============================================================================
  // PAYMENT LIFECYCLE HOOKS
  // ============================================================================
  //
  // These delegate to the underlying BatchEvmScheme so that the `pay()` buyer
  // flow runs the same lifecycle hooks as the standalone scheme / x402Client
  // paths — see https://docs.x402.org/advanced-concepts/lifecycle-hooks.
  /**
   * Register a hook that runs before the payment payload is created. Returning
   * `{ abort: true, reason }` causes `pay()` to throw before signing.
   */
  onBeforePaymentCreation(hook) {
    this.batchScheme.onBeforePaymentCreation(hook);
    return this;
  }
  /**
   * Register a hook that runs after the payment payload is successfully created.
   */
  onAfterPaymentCreation(hook) {
    this.batchScheme.onAfterPaymentCreation(hook);
    return this;
  }
  /**
   * Register a hook that runs when payload creation throws. Return
   * `{ recovered: true, payload }` to recover with the provided payload.
   */
  onPaymentCreationFailure(hook) {
    this.batchScheme.onPaymentCreationFailure(hook);
    return this;
  }
  /**
   * Register a hook that runs after the buyer's paid HTTP request completes.
   * Return `{ recovered: true }` to have `pay()` retry once with a fresh payload.
   */
  onPaymentResponse(hook) {
    this.batchScheme.onPaymentResponse(hook);
    return this;
  }
  /**
   * Get the account address.
   */
  get address() {
    return this.account.address;
  }
  /**
   * Get the chain name.
   */
  get chainName() {
    return this.chainConfig.chain.name;
  }
  /**
   * Get the Gateway domain for this chain.
   */
  get domain() {
    return this.chainConfig.domain;
  }
  /**
   * Get the chain name key for this client.
   */
  getChainName() {
    for (const [name, config] of Object.entries(CHAIN_CONFIGS)) {
      if (config.domain === this.chainConfig.domain && config.chain.id === this.chainConfig.chain.id) {
        return name;
      }
    }
    throw new Error("Unknown chain configuration");
  }
  /**
   * Get the USDC balance of the account (not in Gateway).
   *
   * @param address - Optional address to check (defaults to account address)
   * @returns USDC balance in atomic units and formatted
   */
  async getUsdcBalance(address) {
    const target = address ?? this.account.address;
    const balance = await this.publicClient.readContract({
      address: this.chainConfig.usdc,
      abi: erc20Abi,
      functionName: "balanceOf",
      args: [target]
    });
    return {
      balance,
      formatted: formatUnits(balance, 6)
    };
  }
  /**
   * Get all balances (wallet + gateway) in one call.
   *
   * @param address - Optional address to check (defaults to account address)
   * @returns All balances
   */
  async getBalances(address) {
    const target = address ?? this.account.address;
    const [wallet, gateway] = await Promise.all([
      this.getUsdcBalance(target),
      this.getGatewayBalance(target)
    ]);
    return {
      wallet,
      gateway
    };
  }
  /**
   * Get the Gateway balance for an address.
   *
   * @param address - Optional address to check (defaults to account address)
   * @returns Gateway balance information
   * @deprecated Use `getBalances()` instead for a unified view
   */
  async getBalance(address) {
    return this.getGatewayBalance(address);
  }
  /**
   * Get the Gateway balance for an address via Gateway API.
   */
  async getGatewayBalance(address) {
    const target = address ?? this.account.address;
    const apiBaseUrl = this.isTestnet() ? GATEWAY_API_TESTNET : GATEWAY_API_MAINNET;
    const response = await fetch(`${apiBaseUrl}/balances`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...this.gatewayApiHeaders()
      },
      body: JSON.stringify({
        token: "USDC",
        sources: [{ depositor: target, domain: this.chainConfig.domain }]
      })
    });
    const data = await response.json();
    if (!response.ok) {
      throw new Error(
        `Gateway API balance fetch failed: ${data.message ?? response.statusText}`
      );
    }
    if (!data.balances || data.balances.length === 0) {
      throw new Error("Gateway API returned no balances for the depositor.");
    }
    const balanceData = data.balances[0];
    const available = parseUnits(balanceData.balance, 6);
    const withdrawing = parseUnits(balanceData.withdrawing ?? "0", 6);
    const withdrawable = parseUnits(balanceData.withdrawable ?? "0", 6);
    const total = available + withdrawing;
    return {
      total,
      available,
      withdrawing,
      withdrawable,
      formattedTotal: formatUnits(total, 6),
      formattedAvailable: formatUnits(available, 6),
      formattedWithdrawing: formatUnits(withdrawing, 6),
      formattedWithdrawable: formatUnits(withdrawable, 6)
    };
  }
  /**
   * Deposit USDC into the Gateway Wallet.
   *
   * This method first approves the Gateway Wallet contract to spend USDC,
   * then deposits the specified amount.
   *
   * @param amount - Amount of USDC to deposit (as a decimal string, e.g., "10.5")
   * @param options - Optional deposit options
   * @returns Deposit result with transaction hashes
   */
  async deposit(amount, options) {
    const depositAmount = parseUnits(amount, 6);
    const approveAmount = parseUnits(options?.approveAmount ?? amount, 6);
    const { balance } = await this.getUsdcBalance();
    if (balance < depositAmount) {
      throw new Error(
        `Insufficient USDC balance. Have: ${formatUnits(balance, 6)}, Need: ${amount}`
      );
    }
    let approvalTxHash;
    if (!options?.skipApprovalCheck) {
      const allowance = await this.publicClient.readContract({
        address: this.chainConfig.usdc,
        abi: erc20Abi,
        functionName: "allowance",
        args: [this.account.address, this.chainConfig.gatewayWallet]
      });
      if (allowance < depositAmount) {
        approvalTxHash = await this.walletClient.writeContract({
          address: this.chainConfig.usdc,
          abi: erc20Abi,
          functionName: "approve",
          args: [this.chainConfig.gatewayWallet, approveAmount]
        });
        try {
          await this.publicClient.waitForTransactionReceipt({ hash: approvalTxHash });
        } catch (err) {
          throw new Error(`Approval transaction failed: ${approvalTxHash}`, {
            cause: err
          });
        }
      }
    }
    const depositTxHash = await this.walletClient.writeContract({
      address: this.chainConfig.gatewayWallet,
      abi: GATEWAY_WALLET_ABI,
      functionName: "deposit",
      args: [this.chainConfig.usdc, depositAmount],
      gas: 120000n
    });
    try {
      await this.publicClient.waitForTransactionReceipt({ hash: depositTxHash });
    } catch (err) {
      throw new Error(`Deposit transaction failed: ${depositTxHash}`, {
        cause: err
      });
    }
    return {
      approvalTxHash,
      depositTxHash,
      amount: depositAmount,
      formattedAmount: amount,
      depositor: this.account.address
    };
  }
  /**
   * Deposit USDC into the Gateway Wallet on behalf of another address.
   *
   * The resulting balance belongs to the specified depositor address,
   * not the caller.
   *
   * @param amount - Amount of USDC to deposit (as a decimal string)
   * @param depositor - Address that will own the resulting balance
   * @param options - Optional deposit options
   * @returns Deposit result with transaction hashes
   */
  async depositFor(amount, depositor, options) {
    const depositAmount = parseUnits(amount, 6);
    const approveAmount = parseUnits(options?.approveAmount ?? amount, 6);
    const { balance } = await this.getUsdcBalance();
    if (balance < depositAmount) {
      throw new Error(
        `Insufficient USDC balance. Have: ${formatUnits(balance, 6)}, Need: ${amount}`
      );
    }
    let approvalTxHash;
    if (!options?.skipApprovalCheck) {
      const allowance = await this.publicClient.readContract({
        address: this.chainConfig.usdc,
        abi: erc20Abi,
        functionName: "allowance",
        args: [this.account.address, this.chainConfig.gatewayWallet]
      });
      if (allowance < depositAmount) {
        approvalTxHash = await this.walletClient.writeContract({
          address: this.chainConfig.usdc,
          abi: erc20Abi,
          functionName: "approve",
          args: [this.chainConfig.gatewayWallet, approveAmount]
        });
        try {
          await this.publicClient.waitForTransactionReceipt({ hash: approvalTxHash });
        } catch (err) {
          throw new Error(`Approval transaction failed: ${approvalTxHash}`, {
            cause: err
          });
        }
      }
    }
    const depositTxHash = await this.walletClient.writeContract({
      address: this.chainConfig.gatewayWallet,
      abi: GATEWAY_WALLET_ABI,
      functionName: "depositFor",
      args: [this.chainConfig.usdc, depositor, depositAmount],
      gas: 120000n
    });
    try {
      await this.publicClient.waitForTransactionReceipt({ hash: depositTxHash });
    } catch (err) {
      throw new Error(`Deposit transaction failed: ${depositTxHash}`, {
        cause: err
      });
    }
    return {
      approvalTxHash,
      depositTxHash,
      amount: depositAmount,
      formattedAmount: amount,
      depositor
    };
  }
  // ============================================================================
  // PAYMENT (Pay for x402 resources)
  // ============================================================================
  /**
   * Check if a URL supports Gateway batching before paying.
   *
   * @param url - The URL to check
   * @returns Whether batching is supported and payment requirements
   */
  async supports(url) {
    try {
      const response = await fetch(url);
      if (response.status !== 402) {
        return { supported: false, error: "Resource does not require payment (not 402)" };
      }
      const paymentRequiredHeader = response.headers.get("PAYMENT-REQUIRED");
      if (!paymentRequiredHeader) {
        return {
          supported: false,
          error: "Missing PAYMENT-REQUIRED header in 402 response"
        };
      }
      const data = JSON.parse(
        Buffer.from(paymentRequiredHeader, "base64").toString("utf-8")
      );
      const accepts = data.accepts;
      if (!accepts || accepts.length === 0) {
        return { supported: false, error: "No payment options in 402 response" };
      }
      const expectedNetwork = `eip155:${this.chainConfig.chain.id}`;
      const batchingOption = accepts.find((opt) => {
        const extra = opt.extra;
        return opt.network === expectedNetwork && extra?.name === "GatewayWalletBatched" && extra?.version === "1" && typeof extra?.verifyingContract === "string";
      });
      if (!batchingOption) {
        return {
          supported: false,
          error: `No Gateway batching option available for network ${expectedNetwork} (${this.chainConfig.chain.name})`
        };
      }
      return { supported: true, requirements: batchingOption };
    } catch (error) {
      return { supported: false, error: error.message };
    }
  }
  /**
   * Pay for an x402-protected resource.
   *
   * This method handles the full 402 payment flow automatically:
   * 1. Makes initial request
   * 2. If 402, finds Gateway batching option
   * 3. Signs payment authorization
   * 4. Retries with Payment-Signature header
   *
   * @param url - The URL to pay for
   * @param options - Optional request options
   * @returns The response data and payment info
   *
   * @example
   * ```typescript
   * const { data, amount } = await gateway.pay('https://api.example.com/resource');
   * console.log('Paid', amount, 'USDC for:', data);
   * ```
   */
  async pay(url, options) {
    const method = options?.method ?? "GET";
    const headers = {
      "Content-Type": "application/json",
      ...options?.headers
    };
    const serializedBody = options?.body !== void 0 ? typeof options.body === "string" ? options.body : JSON.stringify(options.body) : void 0;
    const initialResponse = await fetch(url, {
      method,
      headers,
      body: serializedBody
    });
    if (initialResponse.status !== 402) {
      if (initialResponse.ok) {
        const data = await initialResponse.json();
        return {
          data,
          amount: 0n,
          formattedAmount: "0",
          transaction: "",
          status: initialResponse.status
        };
      }
      throw new Error(`Request failed with status ${initialResponse.status}`);
    }
    const paymentRequiredHeader = initialResponse.headers.get("PAYMENT-REQUIRED");
    if (!paymentRequiredHeader) {
      throw new Error("Missing PAYMENT-REQUIRED header in 402 response");
    }
    const paymentRequired = JSON.parse(
      Buffer.from(paymentRequiredHeader, "base64").toString("utf-8")
    );
    const accepts = paymentRequired.accepts;
    if (!accepts || accepts.length === 0) {
      throw new Error("No payment options in 402 response");
    }
    const expectedNetwork = `eip155:${this.chainConfig.chain.id}`;
    const batchingOption = accepts.find((opt) => {
      const extra = opt.extra;
      return opt.network === expectedNetwork && extra?.name === "GatewayWalletBatched" && extra?.version === "1" && typeof extra?.verifyingContract === "string";
    });
    if (!batchingOption) {
      throw new Error(
        `No Gateway batching option available for network ${expectedNetwork} (${this.chainConfig.chain.name}). The seller may not support this chain. Use supports() to check first.`
      );
    }
    const x402Version = paymentRequired.x402Version ?? 2;
    const amount = BigInt(batchingOption.amount);
    for (let attempt = 0; ; attempt++) {
      const paymentPayload = await this.batchScheme.createPaymentPayload(
        x402Version,
        batchingOption
      );
      const paymentHeader = Buffer.from(
        JSON.stringify({
          ...paymentPayload,
          resource: paymentRequired.resource,
          accepted: batchingOption
        })
      ).toString("base64");
      const paidResponse = await fetch(url, {
        method,
        headers: {
          ...headers,
          "Payment-Signature": paymentHeader
        },
        body: serializedBody
      });
      let settleResponse;
      const paymentResponseHeader = paidResponse.headers.get("PAYMENT-RESPONSE");
      if (paymentResponseHeader) {
        settleResponse = JSON.parse(
          Buffer.from(paymentResponseHeader, "base64").toString("utf-8")
        );
      }
      const responseContext = {
        paymentPayload: {
          x402Version: paymentPayload.x402Version,
          payload: paymentPayload.payload
        },
        requirements: batchingOption,
        paymentRequired,
        settleResponse,
        error: paidResponse.ok ? void 0 : new Error(`Payment failed with status ${paidResponse.status}`)
      };
      const recovery = await this.batchScheme.dispatchPaymentResponse(responseContext);
      if (recovery?.recovered && attempt === 0) {
        continue;
      }
      if (!paidResponse.ok) {
        const error = await paidResponse.json().catch(() => ({}));
        throw new Error(
          `Payment failed: ${error.error || paidResponse.statusText}`
        );
      }
      const data = await paidResponse.json();
      return {
        data,
        amount,
        formattedAmount: formatUnits(amount, 6),
        transaction: settleResponse?.transaction ?? "",
        status: paidResponse.status
      };
    }
  }
  // ============================================================================
  // WITHDRAW (Instant way to get USDC out)
  // ============================================================================
  /**
   * Withdraw USDC from Gateway to your wallet.
   *
   * By default, withdraws to the same chain (instant, no 7-day delay).
   * Optionally, withdraw to a different chain (requires gas on destination).
   *
   * @param amount - Amount of USDC to withdraw (as a decimal string)
   * @param options - Optional withdrawal options
   * @returns Withdrawal result with transaction hash
   *
   * @example
   * ```typescript
   * // Withdraw to same chain (instant!)
   * await gateway.withdraw('50');
   *
   * // Withdraw to Base Sepolia (requires ETH on Base for gas)
   * await gateway.withdraw('25', { chain: 'baseSepolia' });
   * ```
   */
  async withdraw(amount, options) {
    const destinationChainName = options?.chain ?? this.getChainName();
    const destConfig = CHAIN_CONFIGS[destinationChainName];
    if (!destConfig) {
      throw new Error(`Unsupported destination chain: ${destinationChainName}`);
    }
    const withdrawAmount = parseUnits(amount, 6);
    const recipient = options?.recipient ?? this.account.address;
    const maxFee = parseUnits(options?.maxFee ?? "2.01", 6);
    const balances = await this.getBalances();
    if (balances.gateway.available < withdrawAmount) {
      throw new Error(
        `Insufficient available balance. Have: ${balances.gateway.formattedAvailable}, Need: ${amount}`
      );
    }
    const burnIntent = this.createBurnIntent(
      this.chainConfig,
      destConfig,
      withdrawAmount,
      recipient,
      maxFee
    );
    const signature = await this.account.signTypedData({
      domain: { name: "GatewayWallet", version: "1" },
      types: {
        EIP712Domain: [
          { name: "name", type: "string" },
          { name: "version", type: "string" }
        ],
        TransferSpec: [
          { name: "version", type: "uint32" },
          { name: "sourceDomain", type: "uint32" },
          { name: "destinationDomain", type: "uint32" },
          { name: "sourceContract", type: "bytes32" },
          { name: "destinationContract", type: "bytes32" },
          { name: "sourceToken", type: "bytes32" },
          { name: "destinationToken", type: "bytes32" },
          { name: "sourceDepositor", type: "bytes32" },
          { name: "destinationRecipient", type: "bytes32" },
          { name: "sourceSigner", type: "bytes32" },
          { name: "destinationCaller", type: "bytes32" },
          { name: "value", type: "uint256" },
          { name: "salt", type: "bytes32" },
          { name: "hookData", type: "bytes" }
        ],
        BurnIntent: [
          { name: "maxBlockHeight", type: "uint256" },
          { name: "maxFee", type: "uint256" },
          { name: "spec", type: "TransferSpec" }
        ]
      },
      primaryType: "BurnIntent",
      message: burnIntent
    });
    const apiUrl = this.isTestnet() ? GATEWAY_API_TESTNET : GATEWAY_API_MAINNET;
    const response = await fetch(`${apiUrl}/transfer`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...this.gatewayApiHeaders()
      },
      body: JSON.stringify(
        [{ burnIntent, signature }],
        (_, v) => typeof v === "bigint" ? v.toString() : v
      )
    });
    const result = await response.json();
    if (result.success === false || result.error || !result.attestation || !result.signature) {
      throw new Error(
        `Gateway API error: ${result.message || result.error || JSON.stringify(result)}`
      );
    }
    const destTransport = destConfig.rpcUrl ? http(destConfig.rpcUrl) : http();
    const destWalletClient = createWalletClient({
      account: this.account,
      chain: destConfig.chain,
      transport: destTransport
    });
    const destPublicClient = createPublicClient({
      chain: destConfig.chain,
      transport: destTransport
    });
    const mintTxHash = await destWalletClient.writeContract({
      address: destConfig.gatewayMinter,
      abi: GATEWAY_MINTER_ABI,
      functionName: "gatewayMint",
      args: [result.attestation, result.signature]
    });
    try {
      await destPublicClient.waitForTransactionReceipt({ hash: mintTxHash });
    } catch (err) {
      throw new Error(`Mint transaction failed: ${mintTxHash}`, { cause: err });
    }
    return {
      mintTxHash,
      amount: withdrawAmount,
      formattedAmount: amount,
      sourceChain: this.chainConfig.chain.name,
      destinationChain: destConfig.chain.name,
      recipient
    };
  }
  /**
   * Transfer USDC to any supported chain (alias for withdraw with destination chain).
   * @deprecated Use `withdraw({ chain })` instead
   */
  async transfer(amount, destinationChain, recipient) {
    return this.withdraw(amount, { chain: destinationChain, recipient });
  }
  /**
   * Create a burn intent for a transfer.
   */
  createBurnIntent(from, to, value, recipient, maxFee) {
    const addressToBytes32 = (addr) => pad(addr.toLowerCase(), { size: 32 });
    return {
      maxBlockHeight: maxUint256,
      maxFee,
      spec: {
        version: 1,
        sourceDomain: from.domain,
        destinationDomain: to.domain,
        sourceContract: addressToBytes32(from.gatewayWallet),
        destinationContract: addressToBytes32(to.gatewayMinter),
        sourceToken: addressToBytes32(from.usdc),
        destinationToken: addressToBytes32(to.usdc),
        sourceDepositor: addressToBytes32(this.account.address),
        destinationRecipient: addressToBytes32(recipient),
        sourceSigner: addressToBytes32(this.account.address),
        destinationCaller: addressToBytes32(zeroAddress),
        value,
        salt: `0x${randomBytes(32).toString("hex")}`,
        hookData: "0x"
      }
    };
  }
  // ============================================================================
  // TRANSFER QUERIES
  // ============================================================================
  /**
   * Get a single x402 transfer by ID.
   *
   * @param id - The transfer UUID
   * @returns Transfer object
   */
  async getTransferById(id) {
    const apiBaseUrl = this.isTestnet() ? GATEWAY_API_TESTNET : GATEWAY_API_MAINNET;
    const response = await fetch(
      `${apiBaseUrl}/x402/transfers/${encodeURIComponent(id)}`,
      { method: "GET", headers: this.gatewayApiHeaders() }
    );
    if (!response.ok) {
      const errorText = await response.text().catch(() => response.statusText);
      throw new Error(
        `Gateway getTransferById failed (${response.status}): ${errorText}`
      );
    }
    return await response.json();
  }
  /**
   * Search x402 transfers with optional filters.
   *
   * Supports cursor-based pagination via `pageAfter` / `pageBefore`.
   * When `network` is omitted, defaults to the client's chain (e.g., `eip155:84532`).
   *
   * @param params - Optional transfer search filters
   * @returns Paginated transfer search response
   */
  async searchTransfers(params = {}) {
    const apiBaseUrl = this.isTestnet() ? GATEWAY_API_TESTNET : GATEWAY_API_MAINNET;
    const query = new URLSearchParams();
    if (params.from) query.set("from", params.from);
    if (params.to) query.set("to", params.to);
    const network = params.network ?? `eip155:${this.chainConfig.chain.id}`;
    query.set("network", network);
    if (params.status) query.set("status", params.status);
    if (params.token) query.set("token", params.token);
    if (params.startDate) query.set("startDate", params.startDate);
    if (params.endDate) query.set("endDate", params.endDate);
    if (typeof params.pageSize === "number")
      query.set("pageSize", String(params.pageSize));
    if (params.pageAfter) query.set("pageAfter", params.pageAfter);
    if (params.pageBefore) query.set("pageBefore", params.pageBefore);
    const path = `${apiBaseUrl}/x402/transfers`;
    const queryString = query.toString().replaceAll("%3A", ":");
    const url = queryString ? `${path}?${queryString}` : path;
    const response = await fetch(url, {
      method: "GET",
      headers: this.gatewayApiHeaders()
    });
    if (!response.ok) {
      const errorText = await response.text().catch(() => response.statusText);
      throw new Error(
        `Gateway searchTransfers failed (${response.status}): ${errorText}`
      );
    }
    const data = await response.json();
    const linkHeader = response.headers?.get?.("Link") ?? null;
    const pagination = this.parseLinkHeader(linkHeader);
    if (!pagination) {
      return data;
    }
    return { ...data, pagination };
  }
  // ============================================================================
  // PRIVATE HELPERS
  // ============================================================================
  /**
   * Gateway API headers (Arc mainnet internal preview).
   */
  gatewayApiHeaders() {
    return arcPrivateMainnetHeaders(this._arcPrivateMainnet);
  }
  /**
   * Check if this client is connected to a testnet.
   */
  isTestnet() {
    const testnetChainIds = new Set(
      [
        "arbitrumSepolia",
        "arcTestnet",
        "avalancheFuji",
        "baseSepolia",
        "sepolia",
        "hyperEvmTestnet",
        "optimismSepolia",
        "polygonAmoy",
        "seiAtlantic",
        "sonicTestnet",
        "unichainSepolia",
        "worldChainSepolia"
      ].map((name) => CHAIN_CONFIGS[name].chain.id)
    );
    return testnetChainIds.has(this.chainConfig.chain.id);
  }
  /**
   * Parse pagination from an RFC 5988 Link header.
   * Could be replaced with 'http-link-header' library
   *
   * Example:
   * <https://.../transfers?pageAfter=...>; rel="next", <...>; rel="self"
   */
  parseLinkHeader(linkHeader) {
    if (!linkHeader) {
      return void 0;
    }
    const pagination = {};
    const linkRegex = /<([^>]+)>\s*;\s*rel="([^"]+)"/g;
    let match = linkRegex.exec(linkHeader);
    while (match) {
      const [, url, rel] = match;
      if (rel === "self" || rel === "first" || rel === "prev" || rel === "next") {
        pagination[rel] = url;
      }
      match = linkRegex.exec(linkHeader);
    }
    if (pagination.next) {
      const cursor = this.getQueryParam(pagination.next, "pageAfter");
      if (cursor) pagination.pageAfter = cursor;
    }
    if (pagination.prev) {
      const cursor = this.getQueryParam(pagination.prev, "pageBefore");
      if (cursor) pagination.pageBefore = cursor;
    }
    return Object.keys(pagination).length > 0 ? pagination : void 0;
  }
  getQueryParam(url, key) {
    try {
      return new URL(url).searchParams.get(key) ?? void 0;
    } catch {
      return void 0;
    }
  }
  // ============================================================================
  // TRUSTLESS WITHDRAWAL (Emergency use only)
  // ============================================================================
  /**
   * Get the trustless withdrawal delay in blocks (~7 days).
   *
   * NOTE: This is for the emergency trustless withdrawal flow only.
   * For normal withdrawals, use `transfer()` to the same chain.
   *
   * @returns Number of blocks to wait after initiating trustless withdrawal
   */
  async getTrustlessWithdrawalDelay() {
    return this.publicClient.readContract({
      address: this.chainConfig.gatewayWallet,
      abi: GATEWAY_WALLET_ABI,
      functionName: "withdrawalDelay"
    });
  }
  /**
   * Get the block at which a pending trustless withdrawal becomes available.
   *
   * @param address - Optional address to check (defaults to account address)
   * @returns Block number, or 0n if no pending withdrawal
   */
  async getTrustlessWithdrawalBlock(address) {
    const target = address ?? this.account.address;
    return this.publicClient.readContract({
      address: this.chainConfig.gatewayWallet,
      abi: GATEWAY_WALLET_ABI,
      functionName: "withdrawalBlock",
      args: [this.chainConfig.usdc, target]
    });
  }
  /**
   * Initiate a trustless withdrawal from the Gateway Wallet.
   *
   * ⚠️ **WARNING: This is for emergency use only!**
   *
   * Use this only when Circle's Gateway API is unavailable. For normal
   * withdrawals, use `transfer()` to the same chain - it's instant.
   *
   * After calling this, you must wait ~7 days (`withdrawalDelay` blocks)
   * before calling `completeTrustlessWithdrawal()`.
   *
   * @param amount - Amount of USDC to withdraw (as a decimal string)
   * @returns Withdrawal initiation result
   */
  async initiateTrustlessWithdrawal(amount) {
    const withdrawAmount = parseUnits(amount, 6);
    const balance = await this.getBalance();
    if (balance.available < withdrawAmount) {
      throw new Error(
        `Insufficient available balance. Have: ${balance.formattedAvailable}, Need: ${amount}`
      );
    }
    const txHash = await this.walletClient.writeContract({
      address: this.chainConfig.gatewayWallet,
      abi: GATEWAY_WALLET_ABI,
      functionName: "initiateWithdrawal",
      args: [this.chainConfig.usdc, withdrawAmount],
      gas: 100000n
    });
    try {
      await this.publicClient.waitForTransactionReceipt({ hash: txHash });
    } catch (err) {
      throw new Error(`Trustless withdrawal initiation failed: ${txHash}`, {
        cause: err
      });
    }
    const withdrawalBlock = await this.getTrustlessWithdrawalBlock();
    return {
      txHash,
      amount: withdrawAmount,
      formattedAmount: amount,
      withdrawalBlock
    };
  }
  /**
   * Complete a trustless withdrawal after the ~7-day delay period.
   *
   * ⚠️ **WARNING: This is for emergency use only!**
   *
   * Use this only when Circle's Gateway API is unavailable. For normal
   * withdrawals, use `transfer()` to the same chain - it's instant.
   *
   * @returns Withdrawal result
   */
  async completeTrustlessWithdrawal() {
    const balance = await this.getBalance();
    if (balance.withdrawable === 0n) {
      const withdrawalBlock = await this.getTrustlessWithdrawalBlock();
      const currentBlock = await this.publicClient.getBlockNumber();
      if (withdrawalBlock > 0n && currentBlock < withdrawalBlock) {
        throw new Error(
          `Trustless withdrawal not yet available. Current block: ${currentBlock}, Available at: ${withdrawalBlock}`
        );
      }
      throw new Error(
        "No withdrawable balance. Call initiateTrustlessWithdrawal() first."
      );
    }
    const withdrawAmount = balance.withdrawable;
    const txHash = await this.walletClient.writeContract({
      address: this.chainConfig.gatewayWallet,
      abi: GATEWAY_WALLET_ABI,
      functionName: "withdraw",
      args: [this.chainConfig.usdc],
      gas: 100000n
    });
    try {
      await this.publicClient.waitForTransactionReceipt({ hash: txHash });
    } catch (err) {
      throw new Error(`Trustless withdrawal completion failed: ${txHash}`, {
        cause: err
      });
    }
    return {
      txHash,
      amount: withdrawAmount,
      formattedAmount: formatUnits(withdrawAmount, 6)
    };
  }
};
export {
  ARC_PRIVATE_MAINNET_HEADER,
  BatchEvmScheme,
  CHAIN_CONFIGS,
  CompositeEvmScheme,
  GATEWAY_AUTH_VALIDITY_WINDOW_SECONDS,
  GATEWAY_DOMAINS,
  GatewayClient,
  arcPrivateMainnetHeaders,
  registerBatchScheme
};
