// @usesigillum/cli/dist/lib/cli-client.js
import { GatewayClient } from "@circle-fin/x402-batching/client";
export class SigillumCliError extends Error {
    exitCode;
    phase;
    detail;
    constructor(phase, summary, detail) {
        super(summary);
        this.name = "SigillumCliError";
        this.exitCode = 3;
        this.phase = phase;
        this.detail = detail;
    }
}
export function createSigillumClient(options) {
    const baseUrl = normalizeBaseUrl(options.baseUrl);
    async function quoteAction(envelope) {
        const response = await postJson(`${baseUrl}/api/quote`, envelope);
        if (!response.ok) {
            throw new SigillumCliError("quote", `Quote request failed with HTTP ${response.status}.`);
        }
        return (await response.json());
    }
    async function inspectAction({ envelope, allowDemoConfirm = options.allowDemoConfirm ?? true, }) {
        const quote = await quoteAction(envelope);
        return inspectQuotedAction({
            envelope,
            quote,
            allowDemoConfirm,
        });
    }
    async function inspectQuotedAction({ envelope, quote, allowDemoConfirm = options.allowDemoConfirm ?? true, }) {
        const inspectUrl = `${baseUrl}/api/inspect`;
        const firstInspectResponse = await postJson(inspectUrl, {
            ...envelope,
            action_id: quote.action_id,
            quote_id: quote.quote_id,
        });
        if (firstInspectResponse.ok) {
            const result = (await firstInspectResponse.json());
            return toInspectResult({
                quote,
                result,
                payment: {
                    amount: quote.amount,
                    mode: result.payment?.mode ?? "unknown",
                    rail: result.payment?.rail ?? "unknown",
                    payment_reference: result.payment?.payment_reference,
                    transaction_hash: normalizeTransactionHash(result.payment?.transaction_hash),
                    settlement_status: result.payment?.settlement_status ?? null,
                    settlement_scope: result.payment?.settlement_scope ?? null,
                    settlement_source: result.payment?.settlement_source ?? null,
                    transaction_confirmed_at: result.payment?.transaction_confirmed_at ?? null,
                    batch_reference: result.payment?.batch_reference ?? null,
                },
            });
        }
        const firstInspectBody = (await safeJson(firstInspectResponse));
        if (firstInspectResponse.status !== 402) {
            throw new SigillumCliError("inspect", `Inspect request failed with HTTP ${firstInspectResponse.status}.`, readReason(firstInspectBody) ?? undefined);
        }
        const declaredMode = readHeader(firstInspectResponse, "x-sigillum-payment-mode") ??
            readNestedString(firstInspectBody, "payment", "mode") ??
            "unknown";
        if (declaredMode === "demo") {
            if (!allowDemoConfirm) {
                throw new SigillumCliError("payment", "Server is in demo payment mode.", "Rerun with demo confirmation enabled or switch the server to x402 mode.");
            }
            const demoResponse = await postJson(inspectUrl, {
                ...envelope,
                action_id: quote.action_id,
                quote_id: quote.quote_id,
                payment_confirmed: true,
                payment_proof: "sigillum-cli-demo-confirmation",
            });
            if (!demoResponse.ok) {
                const demoBody = (await safeJson(demoResponse));
                throw new SigillumCliError("payment", `Demo payment confirmation failed with HTTP ${demoResponse.status}.`, readReason(demoBody) ?? undefined);
            }
            const result = (await demoResponse.json());
            return toInspectResult({
                quote,
                result,
                payment: {
                    amount: quote.amount,
                    mode: result.payment?.mode ?? "demo",
                    rail: result.payment?.rail ?? "local-demo",
                    payment_reference: result.payment?.payment_reference,
                    transaction_hash: normalizeTransactionHash(result.payment?.transaction_hash),
                    settlement_status: result.payment?.settlement_status ?? null,
                    settlement_scope: result.payment?.settlement_scope ?? null,
                    settlement_source: result.payment?.settlement_source ?? null,
                    transaction_confirmed_at: result.payment?.transaction_confirmed_at ?? null,
                    batch_reference: result.payment?.batch_reference ?? null,
                    requirement: firstInspectBody?.payment,
                },
            });
        }
        const gateway = await createGatewayClientFromEnv();
        const autoDepositAmount = readEnv("X402_BUYER_AUTO_DEPOSIT_USDC");
        if (autoDepositAmount) {
            await gateway.deposit(autoDepositAmount);
        }
        try {
            const paidResult = (await gateway.pay(inspectUrl, {
                method: "POST",
                body: {
                    ...envelope,
                    action_id: quote.action_id,
                    quote_id: quote.quote_id,
                },
                headers: {
                    "Content-Type": "application/json",
                },
            }));
            return toInspectResult({
                quote,
                result: paidResult.data,
                payment: {
                    amount: quote.amount,
                    mode: paidResult.data.payment?.mode ?? "x402",
                    rail: paidResult.data.payment?.rail ?? "x402",
                    payment_reference: paidResult.data.payment?.payment_reference,
                    transaction_hash: normalizeTransactionHash(paidResult.data.payment?.transaction_hash) ??
                        normalizeTransactionHash(paidResult.transaction),
                    settlement_status: paidResult.data.payment?.settlement_status ?? null,
                    settlement_scope: paidResult.data.payment?.settlement_scope ?? null,
                    settlement_source: paidResult.data.payment?.settlement_source ?? null,
                    transaction_confirmed_at: paidResult.data.payment?.transaction_confirmed_at ?? null,
                    batch_reference: paidResult.data.payment?.batch_reference ?? null,
                    requirement: firstInspectBody?.payment,
                },
            });
        }
        catch (error) {
            if (isLowerNonceError(error)) {
                throw new SigillumCliError("payment", "x402 payment flow failed because the buyer wallet nonce is behind the network state.", "Your buyer wallet likely has a pending or recently submitted transaction. Wait for the previous Arc testnet transaction to settle, or use a fresh funded buyer key, then retry the same command.");
            }
            throw new SigillumCliError("payment", "x402 payment flow failed.", error instanceof Error ? error.message : String(error));
        }
    }
    async function quote(diff) {
        return quoteAction(buildCodeChangeEnvelope(diff, options.agentName));
    }
    async function inspectDiff({ diff, allowDemoConfirm = options.allowDemoConfirm ?? true, repo, branch, commitSha, }) {
        return inspectAction({
            envelope: buildCodeChangeEnvelope(diff, options.agentName, { repo, branch, commitSha }),
            allowDemoConfirm,
        });
    }
    return {
        quote,
        quoteAction,
        inspectDiff,
        inspectAction,
        inspectQuotedAction,
    };
}
export function loadSigillumEnvFiles({ fs, path, searchDirs, }) {
    const directories = Array.from(new Set([...(searchDirs ?? []), process.cwd()].map((dir) => path.resolve(dir))));
    for (const directory of directories) {
        for (const fileName of [".env.local", ".env"]) {
            const filePath = path.join(directory, fileName);
            if (!fs.existsSync(filePath)) {
                continue;
            }
            const contents = fs.readFileSync(filePath, "utf8");
            for (const line of contents.split(/\r?\n/)) {
                const trimmed = line.trim();
                if (!trimmed || trimmed.startsWith("#")) {
                    continue;
                }
                const separator = trimmed.indexOf("=");
                if (separator <= 0) {
                    continue;
                }
                const key = trimmed.slice(0, separator).trim();
                if (!key || process.env[key] !== undefined) {
                    continue;
                }
                let value = trimmed.slice(separator + 1).trim();
                if ((value.startsWith("\"") && value.endsWith("\"")) ||
                    (value.startsWith("'") && value.endsWith("'"))) {
                    value = value.slice(1, -1);
                }
                process.env[key] = value;
            }
        }
    }
}
function toInspectResult({ quote, result, payment, }) {
    return {
        quote,
        payment,
        receipt: result.receipt,
        agent_decision: result.agent_decision,
    };
}
export function buildCodeChangeEnvelope(diff, agentName = "Sigillum CLI", metadata) {
    return {
        agent: {
            name: agentName,
            type: "cli",
        },
        action_type: "code_change",
        action_input: {
            diff,
            ...(metadata?.repo ? { repo: metadata.repo } : {}),
            ...(metadata?.branch ? { branch: metadata.branch } : {}),
            ...(metadata?.commitSha ? { commit_sha: metadata.commitSha } : {}),
        },
    };
}
export function buildDependencyInstallEnvelope(input, agentName = "DependencyInstallAgent") {
    return {
        agent: {
            name: agentName,
            type: "runner",
        },
        action_type: "dependency_install",
        action_input: input,
    };
}
export function buildDeployActionEnvelope(input, agentName = "DeployActionAgent") {
    return {
        agent: {
            name: agentName,
            type: "runner",
        },
        action_type: "deploy_action",
        action_input: input,
    };
}
async function createGatewayClientFromEnv() {
    const privateKey = readEnv("SIGILLUM_BUYER_PRIVATE_KEY", "X402_BUYER_PRIVATE_KEY");
    if (!privateKey) {
        throw new SigillumCliError("payment", "Buyer private key is required for the x402 payment flow.", "Set X402_BUYER_PRIVATE_KEY or SIGILLUM_BUYER_PRIVATE_KEY.");
    }
    const chain = (readEnv("SIGILLUM_BUYER_CHAIN", "X402_NETWORK") || "arcTestnet");
    const rpcUrl = readEnv("SIGILLUM_BUYER_RPC_URL", "X402_RPC_URL");
    return new GatewayClient({
        chain,
        privateKey: privateKey,
        ...(rpcUrl ? { rpcUrl } : {}),
    });
}
function normalizeBaseUrl(value) {
    return value.replace(/\/+$/, "");
}
async function postJson(url, body) {
    try {
        return await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(body),
        });
    }
    catch (error) {
        throw new SigillumCliError("runtime", `Could not reach the Sigillum API at ${url}.`, [
            "Check X402_API_BASE_URL or SIGILLUM_BASE_URL in .env.local.",
            "If the URL is correct, make sure the hosted API is reachable or that your local Next.js app is running.",
            `Underlying error: ${error instanceof Error ? error.message : String(error)}`,
        ].join(" "));
    }
}
async function safeJson(response) {
    try {
        return await response.json();
    }
    catch {
        return null;
    }
}
function readHeader(response, name) {
    return response.headers.get(name);
}
function readReason(body) {
    if (!body || typeof body !== "object") {
        return undefined;
    }
    return body.reason ?? body.message ?? body.error ?? body.payment?.message;
}
function readNestedString(value, outerKey, innerKey) {
    if (!value?.[outerKey]) {
        return undefined;
    }
    const nestedValue = value[outerKey][innerKey];
    return typeof nestedValue === "string" ? nestedValue : undefined;
}
function normalizeTransactionHash(value) {
    return value && /^0x[a-fA-F0-9]{64}$/.test(value) ? value : undefined;
}
function isLowerNonceError(error) {
    if (!(error instanceof Error)) {
        return false;
    }
    return error.message.toLowerCase().includes("nonce provided for the transaction is lower");
}
function readEnv(...keys) {
    for (const key of keys) {
        const value = process.env[key]?.trim();
        if (value) {
            return value;
        }
    }
    return undefined;
}
